pkg/agent/endpoints/workload/handler_test.go

```diff
@@ -1,602 +1,717 @@
-			}
-			// verify subject
-			spiretest.AssertProtoEqual(t,
-				&structpb.Value{
-					Kind: &structpb.Value_StringValue{
-						StringValue: "spiffe://example.org/blog",
-					},
-				}, resp.Claims.Fields["sub"])
-		})
+	manager := &FakeManager{
+		ca:         params.CA,
+		identities: params.Identities,
+		updates:    params.Updates,
 	}
+
+	handler := workload.New(workload.Config{
+		Manager:  manager,
+		Attestor: &FakeAttestor{err: params.AttestErr},
+	})
+
+	unaryInterceptor, streamInterceptor := middleware.Interceptors(
+		middleware.WithLogger(log),
+	)
+
+	server := grpc.NewServer(
+		grpc.UnaryInterceptor(unaryInterceptor),
+		grpc.StreamInterceptor(streamInterceptor),
+	)
+	workloadPB.RegisterSpiffeWorkloadAPIServer(server, handler)
+	socketPath := spiretest.ServeGRPCServerOnTempSocket(t, server)
+	t.Cleanup(func() { server.Stop() })
+
+	// Provide a cancelable context to ensure the stream is always
+	// closed when the test case is done, and also to ensure that
+	// any unexpected blocking call is timed out.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	conn, err := grpc.DialContext(ctx, "unix://"+socketPath, grpc.WithInsecure())
+	require.NoError(t, err)
+	t.Cleanup(func() { conn.Close() })
+
+	fn(ctx, workloadPB.NewSpiffeWorkloadAPIClient(conn))
+
+	cancel()
+	server.GracefulStop()
+
+	assert.Equal(t, 0, manager.Subscribers(), "there should be no more subscribers")
+
+	spiretest.AssertLogs(t, logHook.AllEntries(), params.ExpectLogs)
 }
 
-func (s *HandlerTestSuite) TestStructFromValues() {
-	expected := &structpb.Struct{
-		Fields: map[string]*structpb.Value{
-			"foo": {
-				Kind: &structpb.Value_StringValue{
-					StringValue: "bar",
-				},
-			},
-			"baz": {
-				Kind: &structpb.Value_NumberValue{
-					NumberValue: 3.0,
-				},
-			},
-		},
-	}
+type FakeManager struct {
+	ca          *testca.CA
+	identities  []cache.Identity
+	updates     []*cache.WorkloadUpdate
+	subscribers int32
+}
 
-	actual, err := structFromValues(map[string]interface{}{
-		"foo": "bar",
-		"baz": 3,
-	})
-	s.Require().NoError(err)
-	s.Require().Equal(expected, actual)
+func (m *FakeManager) MatchingIdentities(selectors []*common.Selector) []cache.Identity {
+	return m.identities
 }
 
-func (s *HandlerTestSuite) workloadUpdate() *cache.WorkloadUpdate {
-	svid, key, err := util.LoadSVIDFixture()
-	s.Require().NoError(err)
-	ca, _, err := util.LoadCAFixture()
-	s.Require().NoError(err)
-
-	identity := cache.Identity{
-		SVID:       []*x509.Certificate{svid},
-		PrivateKey: key,
-		Entry: &common.RegistrationEntry{
-			SpiffeId:      "spiffe://example.org/foo",
-			FederatesWith: []string{"spiffe://otherdomain.test"},
-		},
+func (m *FakeManager) FetchJWTSVID(ctx context.Context, spiffeID string, audience []string) (*client.JWTSVID, error) {
+	svid := m.ca.CreateJWTSVID(spiffeid.RequireFromString(spiffeID), audience)
+	return &client.JWTSVID{
+		Token: svid.Marshal(),
+	}, nil
```

Comment:
Maybe there is an opportunity to exercise paths with manager errors here?
That would allow to also test for the return code `codes.Unavailable` when testing `FetchJWTSVID`.
The `FakeManager` struct may incorporate an attribute to specify an error that can be passed through the `testParams`.

Reply:
Sounds good. I'll add a test case for that.